flang/include/flang/Lower/CallInterface.h

```diff
@@ -0,0 +1,261 @@
+//===-- Lower/CallInterface.h -- Procedure call interface ------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Utility that defines fir call interface for procedure both on caller and
+// and callee side and get the related FuncOp.
+// It does not emit any FIR code but for the created mlir::FuncOp, instead it
+// provides back a container of Symbol (callee side)/ActualArgument (caller
+// side) with additional information for each element describing how it must be
+// plugged with the mlir::FuncOp.
+// It handles the fact that hidden arguments may be inserted for the result.
+// while lowering.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef FORTRAN_LOWER_CALLINTERFACE_H
+#define FORTRAN_LOWER_CALLINTERFACE_H
+
+#include "flang/Common/reference.h"
+#include "mlir/IR/Function.h"
+
+namespace Fortran::semantics {
+class Symbol;
+}
+
+namespace Fortran::evaluate {
+struct ProcedureRef;
+class ActualArgument;
+namespace characteristics {
+struct Procedure;
+}
+} // namespace Fortran::evaluate
+
+namespace Fortran::lower {
+class AbstractConverter;
+class SymMap;
+namespace pft {
+struct FunctionLikeUnit;
+}
+
+/// PassedEntityTypes helps abstracting whether CallInterface is mapping
+/// Symbol to mlir::Value (calle side) or ActualArgument to an index inside the
+/// mlir inputs (since it will be up to the CallInterface user to produce the
+/// related mlir::Value).
```

Comment:
On the caller site, this call interface utility is intended to be built before lowering any `ActualArgument`. Hence there are not any `mlir::Value` that it can map to `ActualArgument`s. So instead, it maps the position where the related `mlir::Value` will have to be placed in the inputs.
On the callee side, the utility has access to the first block of the `FuncOp` so it can directly map the `mlir::Value` arguments to the `Symbol`s.
I rephrased to: "an ActualArgument to a position inside the input vector for the CallOp. (caller side. It will be up to the CallInterface user to produce the mlir::Value that will go in this input vector)."

Reply:
Would it simplify things if it were a requirement to process the arguments beforehand on the calling side? I suppose it means some split phase processing.
1) lower arguments
2) figure out the call's characteristics, etc. using (1)
3) cast the actuals if there are type mismatches

Reply:
It is a possibility, I tried to avoid doing this because I think its will eventually make things simpler to know how the argument needs to be passed before lowering it. Take the following example:

```
subroutine foo1(x)
  real, pointer :: x(:, :)
end subroutine

subroutine foo2(x)
  real  :: x(:, :)
end subroutine

subroutine foo2(x)
  real  :: x(100, 100)
end subroutine
 
 real :: array(200, 200)
 pointer :: x(:, :) => array(1:200:2, 1:200:2) 
 call foo1(x) ! here, we just pass the pointer
 ! here (I think, I need to double check), we need to copy the pointed content into a contiguous temp and pass a descriptor over the temp. An alternative ABI would be to accept non-contiguous inputs and have the callee handle it in case it needs to.
 call foo2(x) 
 ! Here, we need to copy the array into a temp and pass a the base address of the temp.
 call foo3(x) 
```
All these actions could be sneaked inside the utility but I feel its cleaner for the bridge to instructions and apply them whatever way the bridge is doing it.
The current PR does not convey such kind of requirements yet, but the idea would be that `PassedEntity` would get a set of requirements such as: "input must made contiguous"...

Reply:
`sub1` -- shouldn't that pass a reference to the pointer?

`sub2` -- should not make a copy (I think) as the slice can be encoded in the descriptor. Now if CONTIGUOUS is added, then all bets are off.

`sub3` -- agreed, the slice must be reified into a contiguous temp

Reply:
Anyway, the canonical lowering of an actual argument expression is to a memory reference. So, it may be possible to lower all the expressions before knowing the function's signature, per se.

The special care and handling cases would be anything truly pass-by-value (the LEN of a CHARACTER return), emboxing a reference, and handling copy-in semantics, as you say. Each argument has to be processed for these in either case.